README.md
# StarIODeviceSetting SDK for iOS Swift

For more information about the API, supported OS, development environment, and supported printers, please refer to the [manual](https://www.star-m.jp/products/s_print/sdk/starprnt_sdk/manual/ios_swift/en/api_stariodevicesetting_library.html).

## Reference
### Online Manual
- [mC-Print2 Online Manual](http://www.star-m.jp/mcprint2-oml.html)
- [mC-Print3 Online Manual](http://www.star-m.jp/mcprint3-oml.html)

## What is SteadyLAN
- [SteadyLAN (English)](https://www.star-m.jp/products/s_print/mcprint3/manual/en/settings/SteadyLAN.htm)
- [SteadyLAN (Japanese)](https://www.star-m.jp/products/s_print/mcprint3/manual/ja/settings/SteadyLAN.htm)

## Note
- SteadyLAN function must be configured to match the operating system (OS) of the device. It cannot be used with an OS different from the printer's settings.
  For more information, please refer to the [manual](https://www.star-m.jp/products/s_print/sdk/starprnt_sdk/manual/ios_swift/en/api_stariodevicesetting_steaylan_setting.html).


## Copyright

Copyright 2020 Star Micronics Co., Ltd. All rights reserved.

CHANGELOG.md
# V1.0.1_20230331

## StarIO
- Added features
  * Added mC-Label3.
  * `SMBluetoothManager` : Added `SMBluetoothSecuritySSPNumericComparison` to `SMBluetoothSecurity`.
- Changed to Embedded Framework from Static Library.

stario (Ver. 2.10.0)  
stariodevicesetting (Ver. 1.0.1)

# V1.0.1

## StarIO
- Added features
  * Added TSP100IV.
- Changed file format from framework to xcframework.
- Supported for iOS simulator running on Apple Silicon.

## StarIODeviceSetting
- Changed file format from framework to xcframework.
- Supported for iOS simulator running on Apple Silicon.

stario (Ver. 2.9.0)  
stariodevicesetting (Ver. 1.0.1)

# V1.0.0

Initial release for GitHub.

stario (Ver. 2.8.2)  
stariodevicesetting (Ver. 1.0.0)

SDK/StarSteadyLANSettingLabs/AppDelegate.swift
//
//  AppDelegate.swift
//  StarSteadyLANSetting
//
//  Created by 2019-131 on 2020/03/10.
//  Copyright © 2020 StarMicronics Co., Ltd. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



SDK/StarSteadyLANSettingLabs/ViewController.swift
//
//  ViewController.swift
//  StarSteadyLANSetting
//
//  Created by 2019-131 on 2020/03/10.
//  Copyright © 2020 StarMicronics Co., Ltd. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource {

    enum CellParamIndex: Int {
        case portName = 0
        case modelName
        case macAddress
    }
    
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var steadyLANSegmentControl: UISegmentedControl!
    
    var cellArray: NSMutableArray!
    
    var selectedIndexPath: IndexPath!
    var portName:     String!
    var modelName:    String!
    var macAddress:   String!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.cellArray = NSMutableArray()
        self.selectedIndexPath = nil
        self.steadyLANSegmentControl.selectedSegmentIndex = 0
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }


    @IBAction func applySetting(_ sender: Any) {
        
        print("Apply remoteConfig Setting")
        
        var commandArray: [UInt8] = []
        
        if steadyLANSegmentControl.selectedSegmentIndex == 1 {
            commandArray = [0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x39, 0x01, 0x01, //set to SteadyLAN(for iOS)
                            0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x70, 0x01, 0x00] //apply setting. Note: The printer is reset to apply setting when writing this command is completed.

            //The settings for other OSs are as follows. But it will not work on iOS devices.
            // For Android
          //commandArray = [0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x39, 0x01, 0x02, //set to SteadyLAN(for Android)
          //                0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x70, 0x01, 0x00] //apply setting. Note: The printer is reset to apply setting when writing this command is completed.
            // For Windows
          //commandArray = [0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x39, 0x01, 0x03, //set to SteadyLAN(for Windows)
          //                0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x70, 0x01, 0x00] //apply setting. Note: The printer is reset to apply setting when writing this command is completed.
        }
        else {
             commandArray = [0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x39, 0x01, 0x00, //set to SteadyLAN(Disable)
                             0x1b, 0x1d, 0x29, 0x4e, 0x03, 0x00, 0x70, 0x01, 0x00] //apply setting. Note: The printer is reset to apply setting when writing this command is completed.
        }
        
        _ = Communication.sendCommands(commandArray, portName: self.portName, portSettings: "", timeout: 10000, completionHandler: { (communicationResult: CommunicationResult) in
            DispatchQueue.main.async {
                self.showSimpleAlert(title: "Communication Result",
                                     message: Communication.getCommunicationResultMessage(communicationResult),
                                     buttonTitle: "OK",
                                     buttonStyle: .cancel)
            }
        })
    }
    
    @IBAction func readSetting(_ sender: Any) {
        
        print("Read remoteConfig Setting")
        
        _ = Communication.confirmSteadyLANSetting(self.portName, portSettings:"", timeout: 10000, completionHandler:
            { (communicationResult: CommunicationResult, message: String) in
                var dialogMessage: String = ""
                
                if communicationResult.result == .success {
                    dialogMessage = message
                }
                else {
                    dialogMessage = Communication.getCommunicationResultMessage(communicationResult)
                }
                
                self.showSimpleAlert(title: "SteadyLAN Setting",
                                     message: dialogMessage,
                                     buttonTitle: "OK",
                                     buttonStyle: .cancel)
        })
    }
    
    @IBAction func searchPrinter(_ sender: Any) {
        
        print("Search Star Printer")
        
        self.cellArray.removeAllObjects()
        
        var searchPrinterResult: [PortInfo]? = nil
        
        do {
            searchPrinterResult = try SMPort.searchPrinter(target: "ALL:") as? [PortInfo]  //ALL
            //searchPrinterResult = try SMPort.searchPrinter(target: "TCP:") as? [PortInfo] //LAN
            //searchPrinterResult = try SMPort.searchPrinter(target: "BT:")  as? [PortInfo] //Bluetooth
            //searchPrinterResult = try SMPort.searchPrinter(target: "BLE:") as? [PortInfo] //Bluetooth Low Energy
        }
        catch {
            // do nothing
        }
        
        guard let portInfoArray: [PortInfo] = searchPrinterResult else {
            self.tableView.reloadData()
            return
        }
        
        for portInfo: PortInfo in portInfoArray {
            self.cellArray.add([portInfo.portName, portInfo.modelName, portInfo.macAddress])
        
        }
        
        self.tableView.reloadData()
    }
    
    
    func showSimpleAlert(title: String?,
                         message: String?,
                         buttonTitle: String?,
                         buttonStyle: UIAlertAction.Style,
                         completion: ((UIAlertController) -> Void)? = nil) {
        let alertController = UIAlertController(title: title,
                                                message: message,
                                                preferredStyle: .alert)
        
        let action = UIAlertAction(title: buttonTitle, style: buttonStyle, handler: nil)
        
        alertController.addAction(action)
        
        self.present(alertController, animated: true, completion: nil)
        
        DispatchQueue.main.async {
            completion?(alertController)
        }
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.cellArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier: String = "UITableViewCellStyleSubtitle"
        
        var cell: UITableViewCell! = self.tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        
        if cell == nil {
            cell = UITableViewCell(style: .subtitle, reuseIdentifier: cellIdentifier)
        }
        
        if cell != nil {
            let cellParam: [String] = self.cellArray[indexPath.row] as! [String]
            
            cell.textLabel!.text = cellParam[CellParamIndex.modelName.rawValue]
            
            if cellParam[CellParamIndex.macAddress.rawValue] == "" {
                cell.detailTextLabel!.text = cellParam[CellParamIndex.portName.rawValue]
            }
            else {
                cell.detailTextLabel!.text = "\(cellParam[CellParamIndex.portName.rawValue]) (\(cellParam[CellParamIndex.macAddress.rawValue]))"
            }
            
            cell      .textLabel!.textColor = UIColor(red: 0.0, green: 0.5, blue: 1.0, alpha: 1.0)
            cell.detailTextLabel!.textColor = UIColor(red: 0.0, green: 0.5, blue: 1.0, alpha: 1.0)
            
            cell.accessoryType = UITableViewCell.AccessoryType.none
            
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        var cell: UITableViewCell!
        
        if self.selectedIndexPath != nil {
            cell = tableView.cellForRow(at: self.selectedIndexPath)
            
            if cell != nil {
                cell.accessoryType = UITableViewCell.AccessoryType.none
            }
        }
        
        cell = tableView.cellForRow(at: indexPath)!
        
        _ = tableView.visibleCells.map{ $0.accessoryType = .none }
        cell.accessoryType = UITableViewCell.AccessoryType.checkmark
        
        self.selectedIndexPath = indexPath
        
        let cellParam: [String] = self.cellArray[self.selectedIndexPath.row] as! [String]
        self.portName   = cellParam[CellParamIndex.portName  .rawValue]
        self.modelName  = cellParam[CellParamIndex.modelName .rawValue]
        self.macAddress = cellParam[CellParamIndex.macAddress.rawValue]

    }

}



SDK/StarSteadyLANSettingLabs/Communication.swift
//
//  Communication.swift
//  StarSteadyLANSetting
//
//  Created by 2019-131 on 2020/03/10.
//  Copyright © 2020 StarMicronics Co., Ltd. All rights reserved.
//

import Foundation

typealias SendCompletionHandler = (_ communicationResult: CommunicationResult) -> Void

typealias SteadyLANSettingCompletionHandler = (_ communicationResult: CommunicationResult, _ steadyLANSetting: String) -> Void

let sm_true:  UInt32 = 1     // SM_TRUE
let sm_false: UInt32 = 0     // SM_FALSE

class Communication {
    
    static func sendCommands(_ commands: [UInt8]!, portName: String!, portSettings: String!, timeout: UInt32, completionHandler: SendCompletionHandler?) -> Bool {
        var result: Result = .errorOpenPort
        var code: Int = SMStarIOResultCodeFailedError
        
        while true {
            var port : SMPort
            
            do {
                // Modify portSettings argument to improve connectivity when continously connecting via some Ethernet/Wireless LAN model.
                // (Refer Readme for details)
                //              port = try SMPort.getPort(portName: portName, portSettings: "(your original portSettings);l1000)", ioTimeoutMillis: timeout)
                port = try SMPort.getPort(portName: portName, portSettings: portSettings, ioTimeoutMillis: timeout)
                
                defer {
                    SMPort.release(port)
                }
                
                // Sleep to avoid a problem which sometimes cannot communicate with Bluetooth.
                // (Refer Readme for details)
                if #available(iOS 11.0, *), portName.uppercased().hasPrefix("BT:") {
                    Thread.sleep(forTimeInterval: 0.2)
                }
                
                var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
                
                result = .errorWritePort
                
                try port.getParsedStatus(starPrinterStatus: &printerStatus, level: 2)
                
                if printerStatus.offline == sm_true {     // Check printer status.
                    break
                }
                
                let startDate: Date = Date()
                
                var total: UInt32 = 0
                
                while total < UInt32(commands.count) {
                    var written: UInt32 = 0
                    
                    try port.write(writeBuffer: commands, offset: total, size: UInt32(commands.count) - total, numberOfBytesWritten: &written)
                    
                    total += written
                    
                    if Date().timeIntervalSince(startDate) >= 30.0 {     // 30000mS!!!
                        break
                    }
                }
                
                if total < UInt32(commands.count) {
                    break
                }
                
                result = .success
                code = SMStarIOResultCodeSuccess
                
                break
            }
            catch let error as NSError {
                code = error.code
                break
            }
        }
        
        completionHandler?(CommunicationResult.init(result, code))
        
        return result == .success
    }
    
    static func confirmSteadyLANSetting(_ portName: String!, portSettings: String!, timeout: UInt32, completionHandler: SteadyLANSettingCompletionHandler?) -> Bool {
        var result: Result = .errorOpenPort
        var code: Int = SMStarIOResultCodeFailedError
        
        var message: String = ""
        
        while true {
            var port: SMPort
            
            do {
                // Modify portSettings argument to improve connectivity when continously connecting via some Ethernet/Wireless LAN model.
                // (Refer Readme for details)
                //              port = try SMPort.getPort(portName: portName, portSettings: "(your original portSettings);l1000)", ioTimeoutMillis: timeout)
                port = try SMPort.getPort(portName: portName, portSettings: portSettings, ioTimeoutMillis: timeout)
                
                defer {
                    SMPort.release(port)
                }
                
                // Sleep to avoid a problem which sometimes cannot communicate with Bluetooth.
                // (Refer Readme for details)
                if #available(iOS 11.0, *), portName.uppercased().hasPrefix("BT:") {
                    Thread.sleep(forTimeInterval: 0.2)
                }
                
                var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
                
                result = .errorWritePort
                
                try port.getParsedStatus(starPrinterStatus: &printerStatus, level:2)
                
                if printerStatus.offline == sm_true {     // Check printer status.
                    break
                }
                
                let startDate: Date = Date()
                
                var total: UInt32 = 0
                
                let commandArray: [UInt8] = [0x1b, 0x1d, 0x29, 0x4e, 0x02, 0x00, 0x49, 0x01] //confirm SteadyLAN setting
                
                while total < UInt32(commandArray.count) {
                    var written: UInt32 = 0
                    
                    try port.write(writeBuffer: commandArray,
                                   offset: total,
                                   size: UInt32(commandArray.count) - total,
                                   numberOfBytesWritten: &written)
                    
                    total += written
                    
                    if Date().timeIntervalSince(startDate) >= 3.0 {     //  3000mS!!!
                        break
                    }
                }
                
                if total < UInt32(commandArray.count) {
                    break
                }
                
                result = .errorReadPort
                
                
                var receivedData: [UInt8] = [UInt8]()
                
                while true {
                    var buffer: [UInt8] = [UInt8](repeating: 0, count: 1024 + 8)
                    
                    if Date().timeIntervalSince(startDate) >= 3.0 {     //  3000mS!!!
                        break
                    }
                    
                    Thread.sleep(forTimeInterval: 0.01)     // Break time.
                    
                    var readLength: UInt32 = 0
                    
                    try port.read(readBuffer: &buffer, offset: 0, size: 1024, numberOfBytesRead: &readLength)
                    
                    if readLength == 0 {
                        continue;
                    }
                    
                    let resizedBuffer = buffer.prefix(Int(readLength)).map{ $0 }
                    receivedData.append(contentsOf: resizedBuffer)
                    
                    //Check the steadyLAN setting value
                    //The following format is transmitted.
                    //  0x1b 0x1d 0x29 0x4e 0x02 0x00 0x49 0x01 [n] 0x0a 0x00
                    //The value of [n] indicates the SteadyLAN setting.
                    //  0x00: Invalid, 0x01: Valid(For iOS), 0x02: Valid(For Android), 0x03: Valid(For Windows)
                    if receivedData.count >= 11 {
                        for i: Int in 0 ..< Int(receivedData.count) {
                            if receivedData[i + 0] == 0x1b &&
                                receivedData[i + 1] == 0x1d &&
                                receivedData[i + 2] == 0x29 &&
                                receivedData[i + 3] == 0x4e &&
                                receivedData[i + 4] == 0x02 &&
                                receivedData[i + 5] == 0x00 &&
                                receivedData[i + 6] == 0x49 &&
                                receivedData[i + 7] == 0x01 &&
                            //  receivedData[i + 8] is stored the steadylan setting value.
                                receivedData[i + 9] == 0x0a &&
                                receivedData[i + 10] == 0x00 {
                                
                                switch (receivedData[i + 8]) {
                                case 0x01:
                                    message = "SteadyLAN(for iOS)."
                                case 0x02:
                                    message = "SteadyLAN(for Android)."
                                case 0x03:
                                    message = "SteadyLAN(for Windows)."
                            //  case 0x00:
                                default:
                                    message = "SteadyLAN(Disable)."
                                }
                                
                                result = .success
                                break
                            }
                        }
                    }
                    
                    if result == .success {
                        break
                    }
                }
                
                if result != .success {
                    break
                }
                
                result = .success
                code = SMStarIOResultCodeSuccess
                
                break
            }
            catch let error as NSError {
                code = error.code
                break
            }
        }
        
        completionHandler?(CommunicationResult.init(result, code), message)
        
        return result == .success
    }
    
    static func getCommunicationResultMessage(_ communicationResult: CommunicationResult) -> String {
        var message: String
        
        switch communicationResult.result {
        case .success:
            message = "Success!"
        case .errorOpenPort:
            message = "Fail to openPort"
        case .errorBeginCheckedBlock:
            message = "Printer is offline (beginCheckedBlock)"
        case .errorEndCheckedBlock:
            message = "Printer is offline (endCheckedBlock)"
        case .errorReadPort:
            message = "Read port error (readPort)"
        case .errorWritePort:
            message = "Write port error (writePort)"
        default:
            message = "Unknown error"
        }
        
        if communicationResult.result != .success {
            message += "\n\nError code: " + String(communicationResult.code)
            
            if communicationResult.code == SMStarIOResultCodeInUseError {
                message += " (In use)"
            }
            else if communicationResult.code == SMStarIOResultCodeFailedError {
                message += " (Failed)"
            }
        }
        
        return message
    }
}
